### ISSUE TITLE
(aws-ecs-patterns): unable to set DesiredCount to zero

### ISSUE THREAD

**Issue by adrian-baker** [aws/aws-cdk#16562]
The ApplicationLoadBalancedFargateService construct rejects an attempt to set a DesiredCount of zero, however this is valid and supported in the underlying `AWS::ECS::Service` resource.

Common use cases for this:
- I want to "suspend" a service, particularly if it's a headless consumer with background worker tasks.
- During development of a new cluster with a new image, I want to deploy the cluster and supporting infrastructure without attempting to start the Docker image, because I don't want a slow startup attempt, with multiple retries and a lengthly rollback of the entire cluster every time I debug an issue with the image starting up.

### Environment

  - **CDK CLI Version  :** 2.0.0-rc.21 (build f35f136)

### Other

--- 

This is :bug: Bug Report

**Comment by ryparker** [comment]
Hey @adrian-baker 👋🏻 

A similar issue here: https://github.com/aws/aws-cdk/issues/4719

> > @pkandasamy91 It's also currently impossible to set desiredCount: 0 for ALB/NLB EC2/Fargate services. I've changed all 3 in my PR, but we could also add a warning for the SQS queues?
>
> That's intentional, most Load Balanced services scale on memory or cpu usage (which won't apply if 0 tasks are started). I wouldn't recommend providing a desired count of 0 for load balanced services.

However I do understand your use case. I think it warrants a discussion on how opinionated we should be with L3s.

For now though, you could workaround this using an [escape hatch](https://docs.aws.amazon.com/cdk/latest/guide/cfn_layer.html).

e.g.
```ts
const loadBalancedEcsService = new ecsPatterns.ApplicationLoadBalancedEc2Service(stack, 'Service', {
  cluster,
  memoryLimitMiB: 1024,
  taskImageOptions: {
    image: ecs.ContainerImage.fromRegistry('test'),
    environment: {
      TEST_ENVIRONMENT_VARIABLE1: "test environment variable 1 value",
      TEST_ENVIRONMENT_VARIABLE2: "test environment variable 2 value"
    },
  },
  // desiredCount: 0, // CDK throws error when set to 0. Defaults to 1. Overriding with escape hatch.
});

const cfnEcsService = loadBalancedEcsService.service.node.defaultChild as ecs.CfnService
cfnEcsService.addPropertyOverride('DesiredCount', 0);
```

**Comment by ryparker** [comment]
Because this is documented functionality and it is working as expected, i'm going to change this to a feature-request.

**Comment by dbartholomae** [comment]
Another +1 from me. In TypeScript, what I would do to keep the original intent while solving the problem is to allow for a special symbol so that no one can set 0 accidentially without reading the documentation:
```ts
import { NO_INSTANCES_AT_ALL } from 'aws-cdk-lib/aws-ecs'

const loadBalancedEcsService = new ecsPatterns.ApplicationLoadBalancedEc2Service(stack, 'Service', {
  memoryLimitMiB: 1024,
  taskImageOptions: {
    image: ecs.ContainerImage.fromRegistry('test'),
  },
  desiredCount: NO_INSTANCES_AT_ALL
});
```
and define the type of `desiredCount` as
```ts
const NO_INSTANCES_AT_ALL = new Symbol('No instances at all')

interface Props {
  // ...
  desiredCount: number | typeof NO_INSTANCES_AT_ALL
}
```
but I'm not sure this works with jsii, as other languages don't support union types.

**Comment by adrian-skybaker** [comment]
>so that no one can set 0 accidentially without reading the documentation:

I'm sceptical this is actually a real problem the API needs to solve. Who is explicitly specifying zero accidentally and being surprised with the result? Having used the underlying resources which permit zero for many years in production, it's never seemed remotely likely.

**Comment by blimmer** [comment]
I tend to agree. Just a regular ol' number would be great for us. One use-case of setting `desiredCount` to `0` is when we're going down for maintenance, e.g., for an Aurora RDS Upgrade that requires downtime. We use CDK to spin down the service, which drops connections from the database before the upgrade.

**Comment by dbartholomae** [comment]
I personally would also prefer to set it to `0`, but this is a solution that would work for everyone and go with the design principles mentioned by the AWS team.

**Comment by LEQADA** [comment]
I'd also prefer 0. 

It'd be clearer to me as a new user than an Enum.

Any plans to implement it in the near future?

**Comment by bonespiked** [comment]
adding my +1 to the idle thread - since '0' is allowed in the resource, 0 should be allowed in the cdk construct.  Another use case is initial deployment of a new service - I set it to 0 so all the other resources can be deployed, then I can focus on any issues with the container starting as a separate task.
